### DISCUSSION TITLE
Kubernetes Server Configuration - Build vs Runtime and Flag vs Env Var Configuration Options

### DISCUSSION THREAD

**Discussion by stevejr** [keycloak/keycloak#33812] [Q&A]
Hi,

We are running Keycloak in Kubernetes and have a question on the above as the documentation doesn't make it 100% clear when Kubernetes, and containers in general, are involved.

We would like to know if supplying an env var to a Keycloak container overrides/supersedes what was supplied at the container image build time. For example, we supply the env var `KC_HTTP_RELATIVE_PATH` to the container at run time and it works but the documentation says this is actually a build time option and should be supplied with the flag `--http-relative-path`. 

This is confusing because other configuration build time options, like `--health-enabled` appear to not work if they are supplied at run time - `KC_HEALTH_ENABLED`.

So, what approach is correct and how do we know if configuration options should be build, runtime or doesn't matter?

Thanks

**Comment by stianst** [comment]
https://www.keycloak.org/server/all-config?f=build

You can use env variables for build options as well, but if you use an optimized image it will be the value used in the image that is effective, and any chance you do at runtime will be ignored.

**Reply by stevejr** [reply]
Thanks @stianst for the above.

When you say `You can use env variables for build options as well` I assume you mean when we build the container image with the `kc.sh build` command? If yes, currently we do not pass any flags or set any of the environment variables at container image build time. 

This is what is confusing me with some of the build options that appear to also work as plain environment variables supplied to the container at runtime. Another environment variable example is `KC_DB` which we set at runtime but not build time and it works.

Should we actually amend our Dockerfile so that all of the configuration items that are marked as build items are supplied and remove the environment variables from our Kubernetes Statefulset definition?
